sbi/inference/nspe/nspe.py

```diff
@@ -0,0 +1,633 @@
+            epoch_start_time = time.time()
+            for batch in train_loader:
+                self.optimizer.zero_grad()
+                # Get batches on current device.
+                theta_batch, x_batch, masks_batch = (
+                    batch[0].to(self._device),
+                    batch[1].to(self._device),
+                    batch[2].to(self._device),
+                )
+
+                train_losses = self._loss(
+                    theta_batch,
+                    x_batch,
+                    masks_batch,
+                    proposal,
+                    calibration_kernel,
+                    force_first_round_loss=force_first_round_loss,
+                )
+
+                train_loss = torch.mean(train_losses)
+
+                train_loss_sum += train_losses.sum().item()
+
+                train_loss.backward()
+                if clip_max_norm is not None:
+                    clip_grad_norm_(
+                        self._neural_net.parameters(), max_norm=clip_max_norm
+                    )
+                self.optimizer.step()
+
+            self.epoch += 1
+
+            train_loss_average = train_loss_sum / (
+                len(train_loader) * train_loader.batch_size  # type: ignore
+            )
+
+            # NOTE: Due to the inherently noisy nature we do instead log a exponential
+            # moving average of the training loss.
+            if len(self._summary["training_loss"]) == 0:
+                self._summary["training_loss"].append(train_loss_average)
+            else:
+                previous_loss = self._summary["training_loss"][-1]
+                self._summary["training_loss"].append(
+                    (1.0 - ema_loss_decay) * previous_loss
+                    + ema_loss_decay * train_loss_average
+                )
+
+            # Calculate validation performance.
+            self._neural_net.eval()
+            val_loss_sum = 0
+
+            with torch.no_grad():
+                for batch in val_loader:
+                    theta_batch, x_batch, masks_batch = (
+                        batch[0].to(self._device),
+                        batch[1].to(self._device),
+                        batch[2].to(self._device),
+                    )
+                    # Take negative loss here to get validation log_prob.
+                    val_losses = self._loss(
+                        theta_batch,
+                        x_batch,
+                        masks_batch,
+                        proposal,
+                        calibration_kernel,
+                        force_first_round_loss=force_first_round_loss,
+                    )
+                    val_loss_sum += val_losses.sum().item()
+
+            # Take mean over all validation samples.
+            val_loss = val_loss_sum / (
+                len(val_loader) * val_loader.batch_size  # type: ignore
+            )
+
+            # NOTE: Due to the inherently noisy nature we do instead log a exponential
```

Comment:
This makes sense, as the default, but then we should alert the user of this somehow? Or maybe separately save the noisy loss and the ema of the loss?

Reply:
For SNPE, SNLE, SNRE, we currently still have the non-smoothed loss history